core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1029,12 +1062,11 @@ private[spark] class TaskSetManager(
       for (tid <- runningTasksSet) {
         val info = taskInfos(tid)
         val index = info.index
-        if (!successful(index) && copiesRunning(index) == 1 && info.timeRunning(time) > threshold &&
-          !speculatableTasks.contains(index)) {
+        if (!successful(index) && copiesRunning(index) == 1 && info.timeRunning(time) > threshold) {
+          addPendingSpeculativeTask(index)
```

Comment:
In this check we ensure 1 task is running (so it's been dequeued). While the 1 task is running, and we mark it for potential speculation, we are not protecting anymore against duplicates. So each of the queues could have duplicate tasks to speculate.

Could this lead to multiple speculative tasks for a single real task? And also, could it lead to a snowball effect where speculative tasks are created off of speculative tasks that are running? Seems like the previous code was protecting against that in `!speculatableTasks.contains(index)`.

Reply:
Yes, you were right, have updated the code to use HashSet instead of ArrayBuffer. Thank you.

Reply:
can you also add a test case that if you have multiple calls to checkSpeculatbleTasks followed by multiple resources offers, you still only submit one speculative task?

Reply:
Have added the test case.